### ISSUE TITLE
Generating ViewState for typed View

### ISSUE THREAD

**Issue by valeragit** [Arello-Mobile/Moxy#36]
I have an error, when trying to build project.
The problem is with generated ViewState's

I have View:

```
interface LceView<in M : Any> : MvpView {

    fun showLoading(pullToRefresh: Boolean = false)

    fun showContent(model: M)

    fun showError(e: Throwable, pullToRefresh: Boolean = false)
}
```

But generated LceViewState generated with [ ] around Model template:

```
public class LceView$$State<M> extends MvpViewState<LceView> implements LceView<[M]> {
    private ViewCommands<LceView> mViewCommands = new ViewCommands<>();

    @Override
    public void restoreState(LceView view) {
        if (mViewCommands.isEmpty()) {
            return;
        }

        mViewCommands.reapply(view);
    }

    @Override
    public  void showLoading(boolean pullToRefresh) {
        ShowLoadingCommand showLoadingCommand = new ShowLoadingCommand(pullToRefresh);
        mViewCommands.beforeApply(showLoadingCommand);

        if (mViews == null || mViews.isEmpty()) {
            return;
        }

        for(LceView view : mViews) {
            view.showLoading(pullToRefresh);
        }

        mViewCommands.afterApply(showLoadingCommand);
    }

    @Override
    public  void showContent(M model) {
        ShowContentCommand showContentCommand = new ShowContentCommand(model);
        mViewCommands.beforeApply(showContentCommand);

        if (mViews == null || mViews.isEmpty()) {
            return;
        }

        for(LceView view : mViews) {
            view.showContent(model);
        }

        mViewCommands.afterApply(showContentCommand);
    }

    @Override
    public  void showError(java.lang.Throwable e, boolean pullToRefresh) {
        ShowErrorCommand showErrorCommand = new ShowErrorCommand(e, pullToRefresh);
        mViewCommands.beforeApply(showErrorCommand);

        if (mViews == null || mViews.isEmpty()) {
            return;
        }

        for(LceView view : mViews) {
            view.showError(e, pullToRefresh);
        }

        mViewCommands.afterApply(showErrorCommand);
    }


    private class ShowLoadingCommand extends ViewCommand<LceView> {
        public final boolean pullToRefresh;

        ShowLoadingCommand(boolean pullToRefresh) {
            super("showLoading", com.arellomobile.mvp.viewstate.strategy.AddToEndStrategy.class);
            this.pullToRefresh = pullToRefresh;
        }

        @Override
        public void apply(LceView mvpView) {
            mvpView.showLoading(pullToRefresh);
        }
    }

    private class ShowContentCommand extends ViewCommand<LceView> {
        public final M model;

        ShowContentCommand(M model) {
            super("showContent", com.arellomobile.mvp.viewstate.strategy.AddToEndStrategy.class);
            this.model = model;
        }

        @Override
        public void apply(LceView mvpView) {
            mvpView.showContent(model);
        }
    }

    private class ShowErrorCommand extends ViewCommand<LceView> {
        public final java.lang.Throwable e;
        public final boolean pullToRefresh;

        ShowErrorCommand(java.lang.Throwable e, boolean pullToRefresh) {
            super("showError", com.arellomobile.mvp.viewstate.strategy.AddToEndStrategy.class);
            this.e = e;
            this.pullToRefresh = pullToRefresh;
        }

        @Override
        public void apply(LceView mvpView) {
            mvpView.showError(e, pullToRefresh);
        }
    }
}
```

p.s. I tried use invariant for M, it didn't help

**Comment by valeragit** [comment]
Error:

> Error:17:51:47.150 [ERROR] [system.err] ..\presentation\view\LceView$$State.java:10: error: illegal start of type
> 17:51:47.150 [ERROR] [system.err] public class LceView$$State<M> extends MvpViewState<LceView> implements LceView<[M]> {
> 17:51:47.150 [ERROR] [system.err]                                                                                 ^
>

**Comment by senneco** [comment]
@xanderblinov have you a guess what happened there? I didn't use kotlin 🙈

**Comment by senneco** [comment]
@valeragit could you attach archived project? I don't want make kotlin sample project, but your sample can help to fix this issue.

**Comment by xanderblinov** [comment]
Hi guys!

**Comment by xanderblinov** [comment]
Hi guys! @valeragit could you try to replace kotlin class with java to determine error in kotlin implementattion or not?

**Comment by valeragit** [comment]
I created sample project https://github.com/valeragit/KotlinMoxySample

` could you try to replace kotlin class with java`
I'll try later

**Comment by valeragit** [comment]
I tried switch MainPresenter and MainActivity to plain java, it didn't work. Still can't generate proper LceView$$State class

**Comment by senneco** [comment]
Could you get stacktrace/error message?

**Comment by valeragit** [comment]
```
Error:18:08:12.498 [ERROR] [system.err] E:\projects\KotlinSample\app\build\generated\source\kapt2\debug\ru\andrikeev\android\kotlinsample\presentation\view\LceView$$State.java:10: error: illegal start of type
18:08:12.498 [ERROR] [system.err] public class LceView$$State<M> extends MvpViewState<LceView> implements LceView<[M]> {
18:08:12.498 [ERROR] [system.err]                                                                                 ^
18:08:12.499 [ERROR] [system.err] E:\projects\KotlinSample\app\build\generated\source\kapt2\debug\ru\andrikeev\android\kotlinsample\presentation\view\LceView$$State.java:10: error: '{' expected
18:08:12.499 [ERROR] [system.err] public class LceView$$State<M> extends MvpViewState<LceView> implements LceView<[M]> {
18:08:12.499 [ERROR] [system.err]                                                                                  ^
18:08:12.499 [ERROR] [system.err] E:\projects\KotlinSample\app\build\generated\source\kapt2\debug\ru\andrikeev\android\kotlinsample\presentation\view\LceView$$State.java:10: error: illegal start of type
18:08:12.499 [ERROR] [system.err] public class LceView$$State<M> extends MvpViewState<LceView> implements LceView<[M]> {
18:08:12.500 [ERROR] [system.err]                                                                                   ^
18:08:12.500 [ERROR] [system.err] E:\projects\KotlinSample\app\build\generated\source\kapt2\debug\ru\andrikeev\android\kotlinsample\presentation\view\LceView$$State.java:10: error: <identifier> expected
18:08:12.500 [ERROR] [system.err] public class LceView$$State<M> extends MvpViewState<LceView> implements LceView<[M]> {
18:08:12.500 [ERROR] [system.err]                                                                                    ^
18:08:12.500 [ERROR] [system.err] E:\projects\KotlinSample\app\build\generated\source\kapt2\debug\ru\andrikeev\android\kotlinsample\presentation\view\LceView$$State.java:10: error: ';' expected
18:08:12.500 [ERROR] [system.err] public class LceView$$State<M> extends MvpViewState<LceView> implements LceView<[M]> {
18:08:12.500 [ERROR] [system.err]                                                                                     ^
18:08:12.504 [ERROR] [system.err] 5 errors
18:08:12.510 [ERROR] [org.gradle.BuildExceptionReporter] 
18:08:12.510 [ERROR] [org.gradle.BuildExceptionReporter] FAILURE: Build failed with an exception.
18:08:12.510 [ERROR] [org.gradle.BuildExceptionReporter] 
18:08:12.510 [ERROR] [org.gradle.BuildExceptionReporter] * What went wrong:
18:08:12.510 [ERROR] [org.gradle.BuildExceptionReporter] Execution failed for task ':app:compileDebugJavaWithJavac'.
18:08:12.510 [ERROR] [org.gradle.BuildExceptionReporter] > Compilation failed; see the compiler error output for details.
18:08:12.510 [ERROR] [org.gradle.BuildExceptionReporter] 
18:08:12.510 [ERROR] [org.gradle.BuildExceptionReporter] * Try:
18:08:12.510 [ERROR] [org.gradle.BuildExceptionReporter] Run with --stacktrace option to get the stack trace. 

```

Generated LceView$$State similar to one from start post

**Comment by valeragit** [comment]
It seems that there is a problem of generating ViewState for parameterized View. And problem is not related to Kaitlin.
Simple workaround is to add one more interface that implements parameterized View with necessary type.

**Comment by senneco** [comment]
Oh, yes. Currently Moxy doesn't generate code for typed View. It's old issue, but still is in short list. There several problems to determine View type in annotation processor. We working on it.

You can contribute these feature to Moxy – annotation processor is much fun 😄

**Comment by valeragit** [comment]
I try to figure out how to do it =)

**Comment by senneco** [comment]
This is published in version 1.4.6.
